java/org/apache/catalina/filters/RateLimitFilter.java

```diff
@@ -0,0 +1,230 @@
+     */
+    public static final String PARAM_BUCKET_DURATION = "ratelimit.bucket.duration";
+
+    /**
+     * filter init-param to set the bucket number of requests
+     */
+    public static final String PARAM_BUCKET_REQUESTS = "ratelimit.bucket.requests";
+
+    /**
+     * filter init-param to set the enforce flag
+     */
+    public static final String PARAM_ENFORCE = "ratelimit.enforce";
+
+    /**
+     * filter init-param to set a custom status code if requests per duration exceeded
+     */
+    public static final String PARAM_STATUS_CODE = "ratelimit.status.code";
+
+    /**
+     * filter init-param to set a custom status message if requests per duration exceeded
+     */
+    public static final String PARAM_STATUS_MESSAGE = "ratelimit.status.message";
+
+    TimeBucketCounter bucketCounter;
+
+    private int actualRequests;
+
+    private int bucketRequests = DEFAULT_BUCKET_REQUESTS;
+
+    private int bucketDuration = DEFAULT_BUCKET_DURATION;
+
+    private boolean enforce = DEFAULT_ENFORCE;
+    private int statusCode = DEFAULT_STATUS_CODE;
+
+    private String statusMessage = DEFAULT_STATUS_MESSAGE;
+
+    private transient Log log = LogFactory.getLog(RateLimitFilter.class);
+
+    private static final StringManager sm = StringManager.getManager(RateLimitFilter.class);
+
+    /**
+     * @return the actual maximum allowed requests per time bucket
+     */
+    public int getActualRequests() {
+        return actualRequests;
+    }
+
+    /**
+     * @return the actual duration of a time bucket in milliseconds
+     */
+    public int getActualDurationInSeconds() {
+        return bucketCounter.getActualDuration() / 1000;
+    }
+
+    @Override
+    public void init() throws ServletException {
+
+        FilterConfig config = getFilterConfig();
+
+        String param;
+        param = config.getInitParameter(PARAM_BUCKET_DURATION);
+        if (param != null)
+            bucketDuration = Integer.parseInt(param);
+
+        param = config.getInitParameter(PARAM_BUCKET_REQUESTS);
+        if (param != null)
+            bucketRequests = Integer.parseInt(param);
+
+        param = config.getInitParameter(PARAM_ENFORCE);
+        if (param != null)
+            enforce = Boolean.parseBoolean(param);
+
+        param = config.getInitParameter(PARAM_STATUS_CODE);
+        if (param != null)
+            statusCode = Integer.parseInt(param);
+
+        param = config.getInitParameter(PARAM_STATUS_MESSAGE);
+        if (param != null)
+            statusMessage = param;
+
+        bucketCounter = new TimeBucketCounter(bucketDuration);
+
+        actualRequests = (int) Math.round(bucketCounter.getRatio() * bucketRequests);
+
+        log.info(sm.getString("rateLimitFilter.initialized",
+            super.getFilterName(), bucketRequests, bucketDuration, getActualRequests(),
+            getActualDurationInSeconds(), (!enforce ? "Not " : "") + "enforcing")
+        );
+    }
+
+    /**
```

Comment:
This is useless javadoc. It replicates the generic documentation for `Filter.doFilter` and adds nothing.

If you want to describe how *this* filter works, then by all means, do so. But replicating the superclass/interface documentation is not useful, here.

Reply:
Yeah, well, the IDE graciously offered to autocomplete the @Override method definition with the Javadoc and I accepted :(

I will remove that, but is that the right place to describe how this filter works?  I will look at other filter implementations for guidance.

Reply:
I'd leave the javadoc missing. Document the behavior in the class-level documentation and just ignore the doFilter method. The entire purpose of the Filter class is to host the doFilter method, so ... they are kind of the same thing. There's no reason to document the individual method IMO.

Reply:
Chris - sounds good.  Will do.